File: ozy/installers/file.py
import os

from ozy.installer import Installer
from ozy.utils import download_to_file_obj


class SingleFileInstaller(Installer):
    def __init__(self, name, config):
        super().__init__(name, config, 'url', app_name=name)

    def __str__(self):
        return f'file installer from {self.config("url")}'

    def install(self, to_dir):
        os.makedirs(to_dir)
        url = self.config('url')
        app_path = os.path.join(to_dir, self.config('app_name'))
        with open(app_path, 'wb') as output_file:
            download_to_file_obj(output_file, url)
        os.chmod(app_path, 0o774)

File: src/installers/file.rs
use crate::utils::download_to;

use super::installer::Installer;

use anyhow::{anyhow, Error, Result};
use std::os::unix::fs::PermissionsExt;

pub struct File {
    name: String,
    version: String,
    url: String,
}

impl File {
    pub fn new(
        name: &String,
        version: &str,
        app_config: &serde_yaml::Mapping,
    ) -> Result<Self, Error> {
        let url = match app_config.get("url") {
            Some(serde_yaml::Value::String(url)) => url.clone(),
            _ => {
                return Err(anyhow!("Expected a string url in config for {}", name));
            }
        };

        Ok(File {
            name: name.clone(),
            version: version.to_string(),
            url,
        })
    }
}

impl Installer for File {
    fn install(&self, to_dir: &std::path::Path) -> Result<()> {
        eprintln!("Running {}", self.describe());
        std::fs::create_dir_all(to_dir)?;

        let file = to_dir.join(&self.name);
        download_to(&file, &self.url)?;
        std::fs::set_permissions(file, std::fs::Permissions::from_mode(0o774)).unwrap();

        Ok(())
    }

    fn describe(&self) -> String {
        format!("file installer for {} v{}", self.name, self.version)
    }
}
